src/UserCodeRunner.ts

```diff
@@ -1,482 +1,649 @@
+			ts.ScriptTarget.ESNext,
+			undefined,
+			ts.ScriptKind.TS,
+		);
+
+		const tsFileMap = new Map<string, ts.SourceFile>();
+
+		tsFileMap.set(`${USER_FILE_ALIAS}.ts`, userSourceFile);
+		tsFileMap.set(`${EXECUTION_HARNESS_FILENAME}.ts`, executionSourceFile);
+
+		for (const additionalSourceFile of additionalSourceFiles) {
+			tsFileMap.set(additionalSourceFile.fileName, additionalSourceFile);
+		}
+
+		const jsFileMap = new Map<string, ts.SourceFile>();
+
+		const defaultCompilerHost = ts.createCompilerHost({});
+		const customCompilerHost: ts.CompilerHost = {
+			...defaultCompilerHost,
+			getCurrentDirectory(): string {
+				return '';
+			},
+			getSourceFile: (fileName, languageVersion) => {
+				if (tsFileMap.has(fileName)) {
+					return tsFileMap.get(fileName);
+				} else if (fileName.includes('typescript/lib')) {
+					return defaultCompilerHost.getSourceFile(fileName, languageVersion);
+				}
+				return undefined;
+			},
+			writeFile: (filename, data) => {
+				jsFileMap.set(
+					filename,
+					ts.createSourceFile(filename, data, ts.ScriptTarget.ESNext, undefined, ts.ScriptKind.JS),
+				);
+			},
+			readFile(fileName: string): string | undefined {
+				if (tsFileMap.has(fileName)) {
+					return tsFileMap.get(fileName)!.text;
+				}
+				return defaultCompilerHost.readFile(fileName);
+			},
+			fileExists(fileName: string): boolean {
+				return tsFileMap.has(path.basename(fileName));
+			},
+		};
+
+		const program = ts.createProgram(
+			[...additionalSourceFiles.map(f => f.fileName), EXECUTION_HARNESS_FILENAME],
+			{
+				target: ts.ScriptTarget.ESNext,
+				module: ts.ModuleKind.ES2022,
+				lib: ['lib.esnext.d.ts'],
+				sourceMap: true,
+			},
+			customCompilerHost,
+		);
+
+		const sourceErrors: UserCodeError[] = [];
+		ts.getPreEmitDiagnostics(program).forEach(diagnostic => {
+			if (diagnostic.file) {
+				sourceErrors.push(UserCodeTypeError.new(diagnostic, tsFileMap));
+			}
+		});
+
+		const emitResult = program.emit();
+
+		const sourceMap = await new SourceMapConsumer(jsFileMap.get(`${USER_FILE_ALIAS}.js.map`)!.text);
+
+		for (const diagnostic of emitResult.diagnostics) {
+			if (diagnostic.file) {
+				sourceErrors.push(UserCodeTypeError.new(diagnostic, tsFileMap));
+			}
+		}
+
+		if (sourceErrors.length > 0) {
+			return Result.Err(sourceErrors);
+		}
+
+		return Result.Ok({
+			jsFileMap,
+			tsFileMap,
+			sourceMap,
+		});
+	}
+
+	private static hash(str: string): string {
+		return crypto.createHash('sha1').update(str).digest('base64');
+	}
+
+	public async executeUserCode<ArgsType extends any[], ReturnType = any>(
+		userCode: string,
+		args: ArgsType,
+		outputType: string = 'any',
+		argsTypes: string[] = ['any'],
+		timeout: number = 5000,
+		additionalSourceFiles: ts.SourceFile[] = [],
+		context: vm.Context = vm.createContext(),
+	): Promise<Result<ReturnType, UserCodeError[]>> {
+		const userCodeHash = UserCodeRunner.hash(userCode);
+
+		if (!this.user_file_cache.has(userCodeHash)) {
```

Comment:
Is the user file cache cleared on every run, or does it hang around as long as the service is running?

I'm wondering about using the user code hash as a key - the same expansion logic can be run against multiple mission models, right? And it's possible that it would successfully compile for one of them, and have compilation errors for another. If the cache is re-used across expansion runs, my understanding of this code is that the same expansion logic will be compiled once, and then never again, even if it's run against a different mission model.

Reply:
A related edge case that comes to mind is that a user can submit the same expansion logic for multiple activity types. If the first one typechecks successfully, none of the others will be typechecked at all, even though they would probably fail, since they use different activity types.

I think what I'm arriving at is that the `user_file_cache` is meant to be a cache of the `preProcess` function, so its key should include all of the arguments to `preProcess`.

Reply:
Talking with Dylan last week about it, it seems like everything is cached as long as the service is running. When reading through the code, the LRU system will remove very old runs if the queue is full or if a run is more than  ~21 days old.

I hope the below helps explain things better, This is the most confusing part of the system.

In the expansion process, the user will need to define an [expansion_set](https://github.com/NASA-AMMOS/aerie/wiki/Create-an-Expansion-Set)

When the user does an `expansion run` you tell the run which `expansion_set` to use. Let's say the logic is wrong and you get an error. Then the user will have to go back and either redefine an expansion_set that will work or go back further and update the logic. 

We still don't have a versioning system for authored expansion logic so we are basically saying each logic is the latest version so use that for now in the run.

Reply:
I think the cache is limited to 30 minutes. @mattdailis you reminded me though that I wanted to hash based not only on user code, but all included files and input/output values. I think I forgot about this when implementing.